### PR TITLE
fix #5306 chore(nimbus): Parameterize publish_status with status_next

### LINKED ISSUE #5306: Parameterize publish_status with publish_status_next
To distinguish between going through the publish cycle to update a live experiment and end an experiment, let's add a new field called `publish_status_next` that will be `LIVE` when going through a live cycle and `COMPLETE` when ending an experiment.  This way we can determine which kinto calls to make and which status the experiment will be in when the publish cycle completes successfully.  This should also allow us to remove `is_end_requested` entirely.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-1351)
┆Issue Number: EXP-1351

### PR DESCRIPTION
**Pull request by lmorchard**
Because:

* we want to distinguish between going through the publish cycle to
  update a live experiment and end an experiment

This commit:

* adds a new status_next field indicating the intended end target status
  for the current publishing flow in progress

* removes is_end_requested since status_next=Complete expresses the same

* migration and migration tests for is_end_requested and status_next

* updates GQL types and nimbus-ui frontend to use status_next

### CODE REVIEW

**Comment by lmorchard** [comment]
Opening as draft because it seemed like there was one integration test not yet passing. But it passed locally, so it might be an intermittent failure? Will mark ready for review probably tomorrow if CI goes green

**Comment by lmorchard** [comment]
Oh hey, maybe that failure *was* intermittent?

File: app/experimenter/experiments/api/v5/serializers.py [hunk 1]
```diff
@@ -458,12 +458,18 @@ def validate_secondary_outcomes(self, value):
...
+        )
+        if value not in valid_status_next:
+            choices_str = ", ".join(str(choice) for choice in valid_status_next)
             raise serializers.ValidationError(
-                f"Nimbus Experiment has status '{self.instance.status}', but can only "
-                "be ended when set to 'Live'.",
+                f"Invalid choice for status_next: '{value}' - with status "
+                f"'{self.instance.status}', the only valid choices are "
+                f"'{choices_str}'"
             )
```
Comment (jaredlockhart) [inline app/experimenter/experiments/api/v5/serializers.py:471]: Very nice 👍

File: app/experimenter/experiments/constants/nimbus.py [hunk 2]
```diff
@@ -215,6 +215,12 @@ class PublishStatus(models.TextChoices):
     }
     STATUS_ALLOWS_UPDATE = (Status.DRAFT,)
 
+    # Valid status_next values for given status values
+    VALID_STATUS_NEXT_VALUES = {
+        Status.DRAFT: (None, Status.LIVE),
+        Status.LIVE: (None, Status.LIVE, Status.COMPLETE),
+    }
```
Comment (jaredlockhart) [inline app/experimenter/experiments/constants/nimbus.py:222]: 👩‍🍳

File: app/experimenter/experiments/migrations/0174_switch_to_status_next.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from django.db import migrations, models
+
+from experimenter.experiments.constants import NimbusConstants
+
+
+def convert_is_end_requested_to_status_next(apps, schema_editor):
+    NimbusExperiment = apps.get_model("experiments", "NimbusExperiment")
+    NimbusExperiment.objects.filter(
+        status=NimbusConstants.Status.LIVE, is_end_requested=True
+    ).update(status_next=NimbusConstants.Status.COMPLETE)
```
Comment (jaredlockhart) [inline app/experimenter/experiments/migrations/0174_switch_to_status_next.py:12]: Good idea.  There's more cases though, aren't there?  

- Anything `DRAFT` that isn't `IDLE` should have `status_next=LIVE`
- Anything `LIVE` that isn't `IDLE` and has `is_end_requested=False` should have `status_next=LIVE`
- Anything `LIVE` that isn't `IDLE` and has `is_end_requested=True` should have `status_next=COMPLETE`

File: app/experimenter/experiments/tests/factories/nimbus.py [hunk 4]
```diff
@@ -34,67 +34,76 @@
...
         "publish_status": NimbusExperiment.PublishStatus.WAITING,
     }
     LIVE_IDLE = {
         "status": NimbusExperiment.Status.LIVE,
+        "status_next": None,
         "publish_status": NimbusExperiment.PublishStatus.IDLE,
     }
     LIVE_IDLE_ENROLLING = {
         "status": NimbusExperiment.Status.LIVE,
+        # "status_next": NimbusExperiment.Status.LIVE,
```
Comment (jaredlockhart) [inline app/experimenter/experiments/tests/factories/nimbus.py:67]: Should be `None` here right?

File: app/experimenter/experiments/tests/factories/nimbus.py [hunk 5]
```diff
@@ -34,67 +34,76 @@
...
         "publish_status": NimbusExperiment.PublishStatus.IDLE,
         "is_paused": False,
     }
     LIVE_WAITING_ENROLLING = {
         "status": NimbusExperiment.Status.LIVE,
+        # "status_next": NimbusExperiment.Status.LIVE,
```
Comment (jaredlockhart) [inline app/experimenter/experiments/tests/factories/nimbus.py:73]: This will be live when we do the next ticket but for now it's not used right?

File: app/experimenter/experiments/tests/test_migrations.py [hunk 6]
```diff
@@ -1,48 +1,99 @@
-from django.db import connection
-from django.db.migrations.executor import MigrationExecutor
-from django.test.testcases import TransactionTestCase
+from django.utils.text import slugify
+from django_test_migrations.contrib.unittest_case import MigratorTestCase
```
Comment (jaredlockhart) [inline app/experimenter/experiments/tests/test_migrations.py:2]: The template we had didn't work for this?

**Review by jaredlockhart — COMMENTED** [review]
Very very nice @lmorchard 👍 

Couple little questions but generally everything looks exactly right, thank you!

File: app/experimenter/experiments/tests/test_migrations.py (hunk 6 above)
Comment (lmorchard) [reply app/experimenter/experiments/tests/test_migrations.py:2]: No, I fell down a rabbit hole of needing to access models from the pre-migration and post-migration states because of the change in fields. Things broke down when I tried to use the existing factories and the User model. This `django-test-migrations` package wrapped it all up pretty concisely

Comment (jaredlockhart) [reply app/experimenter/experiments/tests/test_migrations.py:2]: Ah yep makes sense 👍

**Comment by jaredlockhart** [comment]
Oh also I was wondering whether we should track `status_next_old/new` in the changelogs.  It's technically getting tracked in the `experiment_data` field, but I'm wondering whether we'll need to filter on it later....  like to figure out launch/pause/update/end dates.  What do you think?

File: app/experimenter/experiments/migrations/0174_switch_to_status_next.py (hunk 3 above)
Comment (lmorchard) [reply app/experimenter/experiments/migrations/0174_switch_to_status_next.py:12]: Ah yeah, I think I tunnel-visioned on the is_end_requested case, will add these

**Comment by lmorchard** [comment]
Reworked the migration to account for more cases. Added status_next to the changelog model, but did *not* come up with a migration for that. Seemed like a bigger effort for data that probably wouldn't be immediately useful

**Comment by jaredlockhart** [comment]
> Oh hey, maybe that failure _was_ intermittent?

Oh also yeah we do have an intermittent failure in the nimbus integration tests atm: https://github.com/mozilla/experimenter/issues/5415

If you see it again and it's not the same, pls also file it.

File: app/experimenter/experiments/changelog_utils/nimbus.py [hunk 7]
```diff
@@ -56,8 +58,10 @@ def generate_nimbus_changelog(experiment, changed_by, message):
     return NimbusChangeLog.objects.create(
         experiment=experiment,
         old_status=old_status,
+        old_status_next=old_status_next,
         old_publish_status=old_publish_status,
         new_status=experiment.status,
+        new_status_next=experiment.status_next,
```
Comment (jaredlockhart) [inline app/experimenter/experiments/changelog_utils/nimbus.py:64]: 👍 👍 👍 👍 👍

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
Just running through the flows locally and it looks like it gets stuck on launch timeout.

Steps to reproduce
1) Create experiment
2) Approve to launch to RS
3) Wait until timeout happens

Expected:
RS is rolled back
Experiment is in DRAFT/REVIEW(LIVE)

Actual:
RS is rolled back
Experiment is in 
Experiment is DRAFT/WAITING(LIVE)


So something got missed somewhere

**Comment by jaredlockhart** [comment]
Oh interesting, I just filled it out as normal.  Let me try again.

**Comment by lmorchard** [comment]
I see a local experiment with draft->waiting(live) that seems stuck which was generated as a dummy experiment, figured that's issue #5280 though

**Comment by jaredlockhart** [comment]
Huh no, I just got the same, timed out in RS but no update in Experimenter.

I'll record a gif.

**Comment by jaredlockhart** [comment]
Oh no, it just didn't work and then worked on the second pass.  That's even more annoying XD

**Comment by lmorchard** [comment]
Couldn't reproduce the timeout issue, but in trying to do so I *did* find a different issue: Couldn't launch from Preview. Added another valid status_next transition and a test to cover that, pushed a new commit

**Comment by lmorchard** [comment]
Ah ha, rebase reveals a new migration got in ahead of me. Gonna fix that and try CI again
